keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1107 @@
++// in the Kubernetes core API ResourceRequirements.
++message KubernetesResources {
++    // Requests and limits from the Kubernetes container config.
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> requests = 1;
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> limits = 2;
++}
+```
+
+The resources (mounts, devices, CDI devices, Kubernetes resources) in the
+CreateContainer request should be identical to what was (pre-)informed in the
+RunPodSandbox request. If they are different, the CRI runtime may fail the
+container creation, for example because changes cannot be applied after a
+VM-based Pod has been created.
+
+#### UpdateContainerResourcesRequest
+
+The UpdateContainerResourcesRequest message is extended to pass down unmodified
+resource requests from the PodSpec.
+
+```diff
+ message UpdateContainerResourcesRequest {
+     // ID of the container to update.
+     string container_id = 1;
+     // Resource configuration specific to Linux containers.
+     LinuxContainerResources linux = 2;
+     // Resource configuration specific to Windows containers.
+     WindowsContainerResources windows = 3;
+     // Unstructured key-value map holding arbitrary additional information for
+     // container resources updating. This can be used for specifying experimental
+     // resources to update or other options to use when updating the container.
+     map<string, string> annotations = 4;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 5;
+ }
+```
+
+### kubelet
+
+Kubelet code is refactored/modified so that all container resources are known
+before sandbox creation. This mainly consists of preparing all mounts (of all
+containers) early.
+
+Kubelet will be be extended to pass down all mounts, devices, CDI devices and
+the unmodified resource requests and limits to the container runtime in all
+related CRI requests, i.e.  RunPodSandbox, CreateContainer and
+UpdateContainerResources.
+
+For example, take a PodSpec:
+
+```yaml
+apiVersion: v1
+kind: Pod
+...
+spec:
+  containers:
+  - name: cnt-1
+    image: k8s.gcr.io/pause
+    resources:
+      requests:
+        cpu: 1
+        memory: 1G
+        example.com/resource: 1
+      limits:
+        cpu: 2
+        memory: 2G
+        example.com/resource: 1
+    volumeMounts:
+    - mountPath: /my-volume
+      name: my-volume
+  volumes:
+  - name: my-volume
+    emptyDir:
+```
+
+Then kubelet will send the following RunPodSandboxRequest when creating the Pod
+(represented here in yaml format):
+
+```yaml
+RunPodSandboxRequest:
+  config:
+  ...
+    podResources:
+      containers:
+      - name: cnt-1
+        kubernetes_resources:
+          requests:
+            cpu: "1"
+            memory: 1G
+            example.com/resource: "1"
+          limits:
+            cpu: "2"
+            memory: 2G
+            example.com/resource: "1"
+        CDI_devices:
```

Comment:
Question: Should this be `cdi_devices` or `cdiDevices` instead? I'm not sure if there is a preference in terms of the definitions.

Reply:
This is just a yaml representation of the protobuf message. In the .proto the field is `cdi_devices`, in the corresponding go struct it's `CDIDevices` but is json/yaml marshalled as `CDI_devices` (I think), looking at this:
https://github.com/kubernetes/cri-api/blob/5eda686b5b1446f079d074e83f7cf57963fcdb01/pkg/apis/runtime/v1/api.pb.go#L4846

Reply:
...so I think `CDI_devices` is consistent, WDYT?

Reply:
I think we've been inconsistent with how we're naming it though. See:
* https://github.com/kubernetes/kubernetes/blob/2c5105e7b80f97ab2da0d47a6db6875aec38331c/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L1108
* https://github.com/kubernetes/kubernetes/blob/2c5105e7b80f97ab2da0d47a6db6875aec38331c/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto#L93
* https://github.com/kubernetes/kubernetes/blob/2c5105e7b80f97ab2da0d47a6db6875aec38331c/staging/src/k8s.io/kubelet/pkg/apis/dra/v1alpha3/api.proto#L67
* https://github.com/kubernetes/kubernetes/blob/2c5105e7b80f97ab2da0d47a6db6875aec38331c/staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/api.proto#L198

Seems like 3 out of 4 are using `CDIDevices` (with `CDI_Devices` in the proto message).

Reply:
For the record, I don't feel too strongly about this and I'm happy with matching what is already in the CRI **or** `CDIDevices`.

Reply:
This is just an example "output". Maybe the naming details can be sorted out in the implementation/code review?

Reply:
@elezar? Is it ok to close as resolved and not get stuck into this naming detail here?